src/pool.rs

```diff
@@ -302,7 +302,20 @@ impl ConnectionPool {
     /// This also warms up the pool for clients that connect when
     /// the pooler starts up.
     async fn validate(&mut self) -> Result<(), Error> {
-        let mut server_infos = Vec::new();
+        // Reset the server_info cache.
+        self.server_info.store(Arc::new(None));
+
+        if let Err(err) = self.server_info().await {
+            error!("Unable to obtain server_info while validating: {:?}", err);
+        }
+
+        Ok(())
+    }
+
+    /// Requests a server info payload from each shard in the pool, returning the
+    /// first valid received. TODO: We are assuming all servers are identical which
+    /// is not true.
+    async fn fetch_server_info(&self) -> Result<BytesMut, Error> {
```

Comment:
This implementation seems odd. PgCat can be used to proxy multiple servers with different Postgres versions. Most clients rely on server info to determine which protocol version to use and which tables to query (e.g. psql relies on this to figure out which system catalog to query for table info).

Reply:
Very true. In some of the code I removed you can see the comment about "we are assuming all servers are identical which is not true" was there before. The problem is that shards are selected at query time, and shards might differ in version. I'm not sure there's a way around that without requiring all servers to be online and the same version running at pgcat boot time in order for this method to return the correct server info. In the previous impl, it would grab the server info from each shard and then only cache the first server info in the list. Since checking each server in the pool isn't required to be "valid" according to the issue this is resolving, we don't need to check each server.